sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java

```diff
@@ -0,0 +1,170 @@
+    this(
+        restriction,
+        new PartitionRestrictionSplitter(),
+        new PartitionRestrictionClaimer(),
+        new PartitionRestrictionProgressChecker());
+  }
+
+  PartitionRestrictionTracker(
+      PartitionRestriction restriction,
+      PartitionRestrictionSplitter splitter,
+      PartitionRestrictionClaimer claimer,
+      PartitionRestrictionProgressChecker progressChecker) {
+    this.splitter = splitter;
+    this.claimer = claimer;
+    this.restriction = restriction;
+    this.progressChecker = progressChecker;
+    this.timeSupplier = () -> Timestamp.now();
+  }
+
+  @VisibleForTesting
+  public void setTimeSupplier(Supplier<Timestamp> timeSupplier) {
+    this.timeSupplier = timeSupplier;
+  }
+
+  @Override
+  public @Nullable SplitResult<PartitionRestriction> trySplit(double fractionOfRemainder) {
+    final String token =
+        Optional.ofNullable(restriction.getMetadata())
+            .map(PartitionRestrictionMetadata::getPartitionToken)
+            .orElse("");
+    final SplitResult<PartitionRestriction> splitResult =
+        splitter.trySplit(fractionOfRemainder, lastClaimedPosition, restriction);
+    LOG.debug(
+        "["
+            + token
+            + "] Try split "
+            + lastClaimedPosition
+            + ", "
+            + restriction
+            + ": "
+            + splitResult);
+    if (splitResult != null) {
+      PartitionRestriction restrictionFromSplit =
+          Optional.ofNullable(splitResult.getPrimary()).orElse(this.restriction);
+      this.restriction = restrictionFromSplit;
+    }
+    return splitResult;
+  }
+
+  @Override
+  public boolean tryClaim(PartitionPosition position) {
+    final boolean canClaim = claimer.tryClaim(restriction, lastClaimedPosition, position);
+
+    if (canClaim) {
+      this.lastClaimedPosition = position;
+    }
+
+    return canClaim;
+  }
+
+  @Override
+  public Progress getProgress() {
+    final String token =
+        Optional.ofNullable(restriction.getMetadata())
+            .map(PartitionRestrictionMetadata::getPartitionToken)
+            .orElse(null);
+    BigDecimal end;
+    if (restriction.getEndTimestamp().compareTo(Timestamp.MAX_VALUE) == 0) {
+      // When the given end timestamp equals to Timestamp.MAX_VALUE, this means that
+      // the end timestamp is not specified which should be a streaming job. So we
+      // use now() as the end timestamp.
+      end = BigDecimal.valueOf(timeSupplier.get().getSeconds());
+    } else {
+      end = BigDecimal.valueOf(restriction.getEndTimestamp().getSeconds());
+    }
+    final Progress progress = progressChecker.getProgress(restriction, lastClaimedPosition, end);
+    LOG.debug("[" + token + "] Progress is " + progress);
+    return progress;
+  }
+
+  @Override
+  public PartitionRestriction currentRestriction() {
+    return restriction;
+  }
+
+  @Override
+  public void checkDone() throws IllegalStateException {
```

Comment:
I think this is a bit more nuanced, especially for the QUERY_CHANGE_STREAM mode: https://github.com/apache/beam/blob/df67c817184acafc4adbc56a725eed4985053171/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/TimestampRangeTracker.java#L173-L194

For the other modes, I am not sure if this logic is valid. We will have to test it (my fear is that a primary split is trying to finish, but it can't because this throws an exception).

Reply:
You're right. So basically, we call checkDone on the primary restriction right after the split happens right? I added the check that if the mode is QUERY_CHANGE_STREAM, we have to make sure the last claimed timestamp is the very last one in the range.